src/test/lib/PathTest.cpp

```diff
@@ -0,0 +1,94 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    MsQuic Path Unittest
+
+--*/
+
+#include "precomp.h"
+#ifdef QUIC_CLOG
+#include "PathTest.cpp.clog.h"
+#endif
+
+struct PathTestContext {
+    CxPlatEvent ShutdownEvent;
+    MsQuicConnection* Connection {nullptr};
+    CxPlatEvent PeerAddrChangedEvent;
+
+    static QUIC_STATUS ConnCallback(_In_ MsQuicConnection* Conn, _In_opt_ void* Context, _Inout_ QUIC_CONNECTION_EVENT* Event) {
+        PathTestContext* Ctx = static_cast<PathTestContext*>(Context);
+        Ctx->Connection = Conn;
+        if (Event->Type == QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE) {
+            Ctx->Connection = nullptr;
+            Ctx->PeerAddrChangedEvent.Set();
+            Ctx->ShutdownEvent.Set();
+        }
+        else if (Event->Type == QUIC_CONNECTION_EVENT_PEER_ADDRESS_CHANGED) {
+            Ctx->PeerAddrChangedEvent.Set();
+        }
+        return QUIC_STATUS_SUCCESS;
+    }
+};
+
+void
+QuicTestLocalPathChanges(
+    _In_ int Family
+    )
+{
+    MsQuicRegistration Registration;
+    TEST_QUIC_SUCCEEDED(Registration.GetInitStatus());
+
+    MsQuicAlpn Alpn("MsQuicTest");
+
+    MsQuicConfiguration ServerConfiguration(Registration, Alpn, ServerSelfSignedCredConfig);
+    TEST_QUIC_SUCCEEDED(ServerConfiguration.GetInitStatus());
+
+    MsQuicCredentialConfig ClientCredConfig;
+    MsQuicConfiguration ClientConfiguration(Registration, Alpn, ClientCredConfig);
+    TEST_QUIC_SUCCEEDED(ClientConfiguration.GetInitStatus());
+
+    PathTestContext Context;
+    MsQuicAutoAcceptListener Listener(Registration, ServerConfiguration, PathTestContext::ConnCallback, &Context);
+    TEST_QUIC_SUCCEEDED(Listener.GetInitStatus());
+    QUIC_ADDRESS_FAMILY QuicAddrFamily = (Family == 4) ? QUIC_ADDRESS_FAMILY_INET : QUIC_ADDRESS_FAMILY_INET6;
+    QuicAddr ServerLocalAddr(QuicAddrFamily);
+    TEST_QUIC_SUCCEEDED(Listener.Start(Alpn, &ServerLocalAddr.SockAddr));
+    TEST_QUIC_SUCCEEDED(Listener.GetLocalAddr(ServerLocalAddr));
+
+    MsQuicConnection Connection(Registration);
+    TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+
+    TEST_QUIC_SUCCEEDED(Connection.StartLocalhost(ClientConfiguration, ServerLocalAddr));
+    TEST_TRUE(Connection.HandshakeCompleteEvent.WaitTimeout(TestWaitTimeout));
+
+    CxPlatSleep(1000);
```

Comment:
Let's figure out a way to not require sleeps. Try to make everything event based.

Reply:
This one was weird, because the handshake completion event would be triggered before the connection actually was marked as completed. And that was causing the old style of this test to fail. I'll see if I can figure out why that wasn't correct.

Reply:
There's a difference between "handshake complete" and "handshake confirmed" based on which side of the connection you are on. The tuple isn't allowed to change until the handshake is confirmed, which we don't actually expose an event for (because the app doesn't really need to know).

Reply:
So if an event for that isn't exposed, I need the sleep. Thats our option.

Reply:
I believe, you can wait for the server's handshake complete and that will be enough. Otherwise, I'd at least recommend a shorter sleep + poll in a loop, so you don't always wait a whole second. And on macOS, we've even had to increase our poll/wait times because things go so slow at times.